[iOS] Adjust the viewport expansion heuristic to better detect full-page modal containers

whsieh opened the pull request:
#### 5081bf058b8d9343d7f7227b486e02e47e31ed2d
<pre>
[iOS] Adjust the viewport expansion heuristic to better detect full-page modal containers
<a href="https://bugs.webkit.org/show_bug.cgi?id=276849">https://bugs.webkit.org/show_bug.cgi?id=276849</a>
<a href="https://rdar.apple.com/132160090">rdar://132160090</a>

Reviewed by Abrar Rahman Protyasha.

Make some more adjustments to the layout viewport expansion heuristic:

-   When hit-testing to see if there&apos;s a large modal container covering most of the viewport,
    perform an area-based hit-test in the entire viewport. This avoids a false negative in the case
    where the center of the viewport hit-tests to a smaller fixed-position container that does not
    cover the screen.

-   Don&apos;t re-expand to double height after shrinking, to avoid cases where the viewport height might
    thrash too frequently.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateLayoutViewportHeightExpansionTimerFired):

Canonical link: <a href="https://commits.webkit.org/281172@main">https://commits.webkit.org/281172@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/163bc6965c41dc0db546f7579ccc0c4d9226b051

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59005 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38333 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11498 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62646 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9448 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9651 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47719 "Found 1 new test failure: http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6741 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35864 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51048 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28577 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32587 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8452 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54539 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8609 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64344 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2917 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8563 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55040 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2927 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51066 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55143 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13061 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2450 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34162 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35246 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36331 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34992 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -4716,19 +4723,20 @@ static bool selectionIsInsideFixedPositionContainer(LocalFrame& frame)
+                return true;
+        }
+
+        return false;
```
Totally optional, but this portion is a nice candidate to write with `std::ranges::any_of`.

```cpp
return std::ranges::any_of(largeViewportConstrainedElements, [listBasedTestResult = hitTestResult.listBasedTestResult()](auto& element) {
    return listBasedTestResult.contains(element);
});
```

whsieh replied:
Nice! Yeah, I'll use `any_of` here.

(Hm... I'm not so sure about copying the set of hit test result nodes when capturing it, though — I think I'll either use `&listBasedTestResult`, or just `[&]` and `hitTestResult.listBasedTestResult()` inside the lambda).

Thanks for the review!

whsieh committed:
[iOS] Adjust the viewport expansion heuristic to better detect full-page modal containers
https://bugs.webkit.org/show_bug.cgi?id=276849
rdar://132160090

Reviewed by Abrar Rahman Protyasha.

Make some more adjustments to the layout viewport expansion heuristic:

-   When hit-testing to see if there's a large modal container covering most of the viewport,
    perform an area-based hit-test in the entire viewport. This avoids a false negative in the case
    where the center of the viewport hit-tests to a smaller fixed-position container that does not
    cover the screen.

-   Don't re-expand to double height after shrinking, to avoid cases where the viewport height might
    thrash too frequently.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateLayoutViewportHeightExpansionTimerFired):

Canonical link: https://commits.webkit.org/281172@main
